dashboard/app/controllers/policy_compliance_controller.rb

```diff
@@ -18,13 +20,74 @@ def child_account_consent
     user = permission_request.user
     #Update User
     if user.child_account_compliance_state != User::ChildAccountCompliance::PERMISSION_GRANTED
-      user.child_account_compliance_state = User::ChildAccountCompliance::PERMISSION_GRANTED
-      user.child_account_compliance_state_last_updated = DateTime.now.new_offset(0)
+      user.update_child_account_compliance(User::ChildAccountCompliance::PERMISSION_GRANTED)
       user.save!
       parent_email = permission_request.parent_email
       ParentMailer.parent_permission_confirmation(parent_email).deliver_now
     end
     @permission_granted = true
     @permission_granted_date = user.child_account_compliance_state_last_updated
   end
+
+  # POST /policy_compliance/child_account_consent
+  # This URL creates a parental permission request. It is invoked via the
+  # lockout panel or potentially user settings by the child account.
+  def child_account_consent_request
+    # If we already comply, don't suddenly invalid it
+    if current_user.child_account_compliance_state == User::ChildAccountCompliance::PERMISSION_GRANTED
+      redirect_back fallback_location: lockout_path and return
+    end
+
+    # Only allow three unique permission requests per day
+    date = Time.zone.today
+    permission_requests = ParentalPermissionRequest.where(
+      user: current_user,
+      created_at: date.midnight..date.end_of_day
+    ).limit(3).count
+
+    # Create a ParentalPermissionRequest token for user and parent email
+    # When the student 'updates' the parental email, we actually just create a
+    # new request row.
+    permission_request = ParentalPermissionRequest.find_or_initialize_by(
+      user: current_user,
+      parent_email: params.require(:'parent-email')
+    )
+
+    # If we are making a new request but already sent too many today,
+    # just bail and return whence we came
+    if permission_request.new_record? && permission_requests >= 3
+      redirect_back fallback_location: lockout_path and return
+    end
+
+    # If this is not a new row, we are resending the email
+    if permission_request.persisted?
+      permission_request.resends_sent += 1
+    end
```

Comment:
Why isn't this after the `email.deliver_now()` call? I'm imagining we see a row in the database with `resends_sent = 1000` but we only actually sent 3 emails given the below check of `resends_sent`. Maybe all the `.save!` calls should be moved to after the email being sent?

Reply:
This is still before the `save!`, so if the `resends_sent` in the unpersisted record gets to 3, it won't be saved. So there shouldn't be a reason the row would have any number other than 0, 1, or 2. We can certainly move the `save!` after the email save, though. We run the risk of potentially sending the email and not having the corresponding record in some perhaps unlikely event there is an error.

Reply:
If it isn't clear... there is no request record at all until that `save!` for the first time and email is used by a student.

Reply:
Oh that makes sense. I simulated the code in my head incorrectly. Ignore this comment then!